Fix 'ref' parsing crash

RikkiGibson opened the pull request:
Closes #39072

CyrusNajmabadi commented on src/Compilers/CSharp/Portable/Parser/LanguageParser.cs:
```diff
@@ -9223,6 +9223,10 @@ private ExpressionSyntax ParseTerm(Precedence precedence)
                     break;
+                case SyntaxKind.RefKeyword:
+                    // ref is not expected to appear in this position.
+                    expr = this.AddError(ParsePossibleRefExpression(), ErrorCode.ERR_InvalidExprTerm, SyntaxFacts.GetText(tk));
```
it worries me that we still hve infinite loop holes here.  It seems like we're getting into parseterm, but there's no guarantee that parseterm will make progress.  and the caller loops without checking for that.  either ParseTerm should always make progrss (probably a bad idea), or the caller must ensure that if it's not making progress it skips a token/tokens and continues on.

RikkiGibson replied:
It looks like the parser has an `IsMakingProgress` method which is meant to catch this. I will take a look at why it did not catch it in this scenario.

RikkiGibson replied:
Opened #39331 as part of this work

RikkiGibson commented:
Do you think there’s at least a reasonable way to detect that the parser is not making progress and throw?

CyrusNajmabadi commented:
> Do you think there’s at least a reasonable way to detect that the parser is not making progress and throw?

Sure.  In any loop we can just see if we've made no progress after performing the loop body.  i.e. if our current token afterwards is the same as before, then we should just consume a token (into a skipped token) and then proceed with the loop, making sure we attach the skipped token to something.

RikkiGibson commented:
This solution has some unfortunate effects in tests like `SymbolCompletionProviderTests.AfterRefInLambda_TypeOnly`. We don't see `(ref $$` as a type-only context.

CyrusNajmabadi commented:
> This solution has some unfortunate effects in tests like 

We may have to either:
1. figure out a different parser strategy that produces the same trees we used to here.
2. update those features to better support the trees that might show up now.

My preference is the latter.

RikkiGibson commented:
I changed several `while (true)` loops to `while (IsMakingProgress(ref lastTokenPosition))`. For each one I added a test where:
- it would OOM in master
- it would hit the assert once changed to IsMakingProgress.

RikkiGibson commented:
Could I please get a sign-off from @dotnet/roslyn-ide for the ContextQuery changes?

sharwell commented on src/Workspaces/CSharp/Portable/Extensions/ContextQuery/SyntaxTreeExtensions.cs:
```diff
@@ -1559,20 +1559,23 @@ public static bool HasNames(this TupleExpressionSyntax tuple)
+                var parent = token.Parent;
+                if (parent.IsKind(SyntaxKind.RefType) ||
+                    parent.IsKind(SyntaxKind.RefExpression) ||
+                    parent.IsKind(SyntaxKind.LocalDeclarationStatement))
```
💡 Should be able to use the `IsKind` extension that supports multiple kinds

```
parent.IsKind(SyntaxKind.RefType, SyntaxKind.RefExpression, SyntaxKind.LocalDeclarationStatement)
```
